zuko/nn.py

```diff
@@ -270,7 +270,14 @@ def __init__(
         adjacency, inverse = torch.unique(adjacency, dim=0, return_inverse=True)
 
         # P_ij = 1 if A_ik = 1 for all k such that A_jk = 1
-        precedence = adjacency.int() @ adjacency.int().t() == adjacency.sum(dim=-1)
+        # PyTorch doesn't support this operation for integer arrays on CUDA devices
+        precedence = (
+            adjacency.cpu().int() @ adjacency.cpu().int().t() == adjacency.sum(dim=-1).cpu()
+        )
+        try:
+            precedence = precedence.to(torch.get_default_device())
+        except AttributeError:
+            precedence = torch.tensor(precedence.detach().cpu().numpy())
```

Comment:
The cast to `.int()` is here because matmul is not possible for boolean tensors on CPU (and CUDA). We can simply cast to `.double()` instead to support CUDA.